### PR TITLE
More 3D diagnostic tendencies

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
**Note**: This cannot be merged until the submodules for FV3 and FV3/ccpp/physics are changed to NOAA-GSL. They point to SamuelTrahanNOAA so people can test the feature via recursive clone.

CCPP PR: https://github.com/NOAA-GSL/ccpp-physics/pull/81
fv3atm PR: https://github.com/NOAA-GSL/fv3atm/pull/76

-----------------------------------------------

This adds tendencies for all tracers and physics schemes. Tendencies in the prior version match tendencies in this one exactly, except a few tendencies that were erroneously zero before. There are two new regression tests at the ufs-weather-model level for fv3_gfs_v16 and fv3_gsd with tendencies enabled.

The 3D diagnostic tendency variable names look like this:

    dtend_(tracer)_(process)

The namelist can turn on and off memory storage and calculations of the diagnostic tendencies using the dtend_select variable. By default, all variables are calculated. This example would turn off all diagnostic tendencies except microphysics and ozone tendencies:

    &gfs_physics_nml
       ldiag3d = .true.
       qdiag3d = .true.
       lssav = .true.
       dtend_select = "*_mp", "*o3*"
    /

You still have to add the variables to the diag_table:

    "gfs_phys", "dtend_liq_wat_mp", "dtend_liq_wat_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_ice_wat_mp", "dtend_ice_wat_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_rainwat_mp", "dtend_rainwat_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_snowwat_mp", "dtend_snowwat_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_graupel_mp", "dtend_graupel_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_water_nc_mp", "dtend_water_nc_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_ice_nc_mp", "dtend_ice_nc_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_rain_nc_mp", "dtend_rain_nc_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_liq_aero_mp", "dtend_liq_aero_mp", "fv3_history", "all", .false., "none", 2
    "gfs_phys", "dtend_ice_aero_mp", "dtend_ice_aero_mp", "fv3_history", "all", .false., "none", 2
    ...

A full list of possible variables is printed out when you run the model:

    ExtDiag( 197) = dtend(:,:,   4) = dtend_temp_pbl (gfs_phys: temperature tendency due to PBL)
    ExtDiag( 198) = dtend(:,:,  13) = dtend_u_pbl (gfs_phys: x wind tendency due to PBL)
    ExtDiag( 199) = dtend(:,:,  14) = dtend_v_pbl (gfs_phys: y wind tendency due to PBL)
    ....

To avoid needing a terabyte of RAM on each node, variables are stored in a 4D sparse array.  This replaces dt3dt, dq3dt, du3dt, and dv3dt:

    dtend(:,:,dtidx(:,:))

The dtidx turns the outer two sparse dimensions into a non-sparse third dimension of dtend. The two dimensions of dtidx are tracer and tendency. Unused entries are set to 1, and `dtend(:,:,1)` is unused. That way, code that doesn't check if `dtidx(tr,te)>1` won't crash. Tracers in dtidx are accessed like so:

    dtidx(100+ntke,index_of_cause_pbl)

The `100+` means unused tracers, which have a value of `-99` or `0` (depending on the context) will refer to valid locations in the dtidx array. They also free up space for non-tracer tendencies in indices 2-99:

    dtidx(index_for_temperature,:)
    dtidx(index_for_x_wind,:)
    dtidx(index_for_y_wind,:)

I've added tendencies for all tracers in convection, microphysics, convective transport, physics, non-physics, and pbl. Other schemes already had all 3D diagnostic tendencies for all tracers they tend.

### CODE REVIEW

File: tests/parm/v16_c96.nml.IN [hunk 1]
```diff
@@ -120,9 +121,11 @@ deflate_level=1
 /
 
 &gfs_physics_nml
-  fhzero       = 6
+  fhzero       = @[FHZERO]
   h2o_phys     = .true.
-  ldiag3d      = .false.
+  lssav          = @[LDIAG3D]
```
Comment (DomHeinzeller) [inline tests/parm/v16_c96.nml.IN]: if `lssav` needs to switch, then it should have its own variable, not use `LDIAG3D`. However, setting `lssav` is a moot point because it is hardcoded to `.true.` in `GFS_time_vary_pre_timestep_init` (ever since IPD), so please revert this change.

File: tests/rt.conf [hunk 2]
```diff
@@ -112,7 +114,6 @@ RUN     | fv3_regional_control_debug
 RUN     | fv3_control_debug                                                                                                       |                                         | fv3 |
 RUN     | fv3_stretched_nest_debug                                                                                                |                                         | fv3 |
 RUN     | fv3_gsd_debug                                                                                                           |                                         | fv3 |
-RUN     | fv3_gsd_diag3d_debug                                                                                                    |                                         | fv3 |
```
Comment (DomHeinzeller) [inline tests/rt.conf:115]: We need a debug test, choose one of the two new tests that you created.

File: tests/rt_gnu.conf [hunk 3]
```diff
@@ -7,13 +7,15 @@ RUN     | fv3_gfdlmp
...
 RUN     | fv3_gfs_v16_restart                                                                                                     |                |             | fv3_gfs_v16
 RUN     | fv3_gfs_v16_stochy                                                                                                      |                | fv3         |
 RUN     | fv3_gfs_v16_flake                                                                                                       |                | fv3         |
 RUN     | fv3_gfs_v16_RRTMGP                                                                                                      |                | fv3         |
 
 COMPILE | APP=ATM SUITES=FV3_GSD_v0,FV3_GFS_v16_thompson,FV3_RRFS_v1beta 32BIT=Y                                                  |                | fv3         |
 RUN     | fv3_gsd                                                                                                                 |                | fv3         |
+RUN     | fv3_gsd_dtend                                                                                                           |                | fv3         |
```
Comment (DomHeinzeller) [inline tests/rt_gnu.conf]: Same here, add debug test for one of the tests that you created.

File: tests/run_test.sh [hunk 4]
```diff
@@ -113,6 +113,9 @@ fi
 if [[ $DATM = 'true' ]]; then
   cp ${PATHRT}/parm/datm_data_table.IN datm_data_table
 fi
+if [[ "${DIAG_TABLE_ADDITIONAL:-}Q" != Q ]] ; then
```
Comment (DomHeinzeller) [inline tests/run_test.sh:116]: Need to add at least one line of documentation here.

File: tests/tests/fv3_gfs_v16_dtend [hunk 5]
```diff
@@ -0,0 +1,92 @@
+###############################################################################
+#
+#  FV3 CCPP GFS v16 compiled with 32-bit dynamics test in DEBUG mode
```
Comment (DomHeinzeller) [inline tests/tests/fv3_gfs_v16_dtend]: As far as I can see, this test is not run in DEBUG mode, please fix line 3 and 7. Also add information that this test runs with extended/full tendencies.

File: tests/tests/fv3_gsd_dtend [hunk 6]
```diff
@@ -1,12 +1,12 @@
 ###############################################################################
 #
-#  FV3 CCPP GSD (GF CU + Thompson MP + MYNN PBL + RUC LSM) test with 3d tendencies in DEBUG mode
+#  FV3 CCPP GSD (GF CU + Thompson MP + MYNN PBL + RUC LSM) full (0-48h) test
```
Comment (DomHeinzeller) [inline tests/tests/fv3_gsd_dtend]: Pleasee add information that this test runs with extended/full tendencies.

File: tests/parm/diag_additional_gsd_dtend [hunk 7]
```diff
@@ -0,0 +1,120 @@
+# Old d*3dt variables, renamed from their new names:
```
Comment (DomHeinzeller) [inline tests/parm/diag_additional_gsd_dtend:4]: Can you add a one-liner comment about the purpose of this file please?

File: tests/parm/diag_additional_gfs_v16_dtend [hunk 8]
```diff
@@ -0,0 +1,96 @@
+
```
Comment (DomHeinzeller) [inline tests/parm/diag_additional_gfs_v16_dtend:3]: Can you add a one-liner comment about the purpose of this file please?

File: tests/default_vars.sh [hunk 9]
```diff
@@ -442,6 +442,7 @@ export FHROT=0
 export LDIAG3D=.F.
 export QDIAG3D=.F.
 export MAX_OUTPUT_FIELDS=300
+export FHZERO="6"
```
Comment (DomHeinzeller) [inline tests/default_vars.sh:445]: Is this safe? Will it impact any of the existing regression tests/test results? I don't think so, but please confirm.

File: tests/default_vars.sh [hunk 10]
```diff
@@ -483,6 +484,7 @@ export DATM=false
 export DAYS="1"
 export FHMAX="24"
 export FDIAG="6"
+export FHZERO="6"
```
Comment (DomHeinzeller) [inline tests/default_vars.sh:487]: ditto

**Review by DomHeinzeller — CHANGES_REQUESTED** [review]
Please revert `tests/RegressionTests_hera.gnu.log` and `tests/RegressionTests_hera.intel.log`. We don't update these files in gsl/develop, because it just creates merge conflicts when we go back to the authoritative repositories. These tests were run against your own baselines and thus don't tell us what has changed and what hasn't.

You should do the following: after addressing the review comments in this PR and associated PRs (will review in a bit), run tests on hera with Intel and GNU against the existing baselines using `rt.conf` for Intel and `rt_gnu.conf` for GNU. Pipe the output of rt.sh to log files and attach them to the PR as comments. Note which tests fail, make sure that this is expected, and that the tests run to completion and fail in the b4b check.

Next, on Hera with Intel and GNU, create your own temporary baselines using `-c` for `rt_ccpp_dev.conf` (usually, because we add our tests there, but in your case using `rt.conf` and `rt_gnu.conf`), then verify against it using `-m` and attach the logs of each of these `rt.sh` runs to the PR as comment.

File: tests/default_vars.sh (hunk 9 above)
Comment (SamuelTrahanNOAA) [reply tests/default_vars.sh:445]: I looked at the tests, and saw none that would be affected. Also, the regression tests passed on Hera for rt.conf and rt_gnu.conf.

File: tests/default_vars.sh (hunk 10 above)
Comment (SamuelTrahanNOAA) [reply tests/default_vars.sh:487]: I looked at the tests, and saw none that would be affected. Also, the regression tests passed on Hera for rt.conf and rt_gnu.conf.

**Comment by SamuelTrahanNOAA** [comment]
[RegressionTests_hera.intel.log](https://github.com/NOAA-GSL/ufs-weather-model/files/6299879/RegressionTests_hera.intel.log)
[RegressionTests_hera.gnu.log](https://github.com/NOAA-GSL/ufs-weather-model/files/6299880/RegressionTests_hera.gnu.log)

Regression tests passed on Hera for GCC and Intel compiler suites (rt_gnu.conf and rt.conf). There are new baselines here:

 - Hera: /scratch2/BMC/wrfruc/Samuel.Trahan/hydro-tend/FV3_RT/develop-20210330

Only these four directories are updated:

 - GNU/fv3_gfs_v16_dtend_debug
 - GNU/fv3_gsd_dtend_debug
 - INTEL/fv3_gfs_v16_dtend_debug
 - INTEL/fv3_gsd_dtend_debug

File: tests/tests/fv3_gsd_dtend_debug [hunk 11]
```diff
@@ -99,7 +105,5 @@ export LSM=3
 export LSOIL_LSM=9
 export KICE=9
 
-export PRINT_DIFF_PGR=.true.
```
Comment (DomHeinzeller) [inline tests/tests/fv3_gsd_dtend_debug:102]: No need to change this now, but since we remove it from this test we should add it to some other test before our code changes go back to the authoritative repositories. For example the `FV3_GSD_noah` test. The challenge will be to remember doing it.

**Review by DomHeinzeller — APPROVED** [review]

File: tests/tests/fv3_gsd_dtend_debug (hunk 11 above)
Comment (SamuelTrahanNOAA) [reply tests/tests/fv3_gsd_dtend_debug:102]: You could open an issue?

Comment (DomHeinzeller) [reply tests/tests/fv3_gsd_dtend_debug:102]: Done: https://github.com/NOAA-GSL/ufs-weather-model/issues/71
